plugins/lint/src/utils/lintUtils.ts

```diff
@@ -96,7 +98,8 @@ async function lintJS(args: LintArgs): Promise<number> {
     await createESLintAnnotations(report.results);
   }
 
-  return report.errorCount;
+  const returnCode = (report.errorCount > 0 || report.warningCount > maxWarnings) ? 1 : 0;
```

Comment:
As mentioned in https://github.com/eslint/eslint/issues/4125#issuecomment-147839509, the `--max-warnings` functionality isn't handled by `CLIEngine` itself, rather by the `cli.js` by just looking at the total number of warnings and comparing it to the max warnings.